packages/xo-server-auth-ldap/src/index.js

```diff
@@ -250,6 +340,161 @@ class AuthLdap {
+            // Remove it from xoGroups as we will then delete all the remaining
+            // LDAP-provided groups
+            ;[xoGroup] = xoGroups.splice(xoGroupIndex, 1)
+            await this._xo.updateGroup(xoGroup.id, { name })
+          }
+          xoGroup = await this._xo.getGroup(xoGroup.id)
+
+          let ldapGroupMembers = ldapGroup[membersAttribute]
+          ldapGroupMembers = Array.isArray(ldapGroupMembers)
+            ? ldapGroupMembers
+            : [ldapGroupMembers]
+
+          const xoGroupMembers = xoGroup.users.slice(0)
+
+          const actions = []
+          ldapGroupMembers.forEach(ldapId => {
+            const xoUser = xoUsers.find(
+              user => user.authProviders.ldap.id === ldapId
+            )
+            if (xoUser === undefined) {
+              return
+            }
+            // If the user exists in XO, should be a member of the LDAP-provided
+            // group but isn't: add it
+            const userIdIndex = xoGroupMembers.findIndex(id => id === xoUser.id)
+            if (userIdIndex !== -1) {
+              xoGroupMembers.splice(userIdIndex, 1)
+              return
+            }
+
+            actions.push(() => this._xo.addUserToGroup(xoUser.id, xoGroup.id))
+          })
+
+          // All the remaining users of that group can be removed from it since
+          // they're not in the LDAP group
+          actions.push(
+            ...xoGroupMembers.map(userId => () =>
+              this._xo.removeUserFromGroup(userId, xoGroup.id)
+            )
+          )
+          // One by one to avoid race conditions
+          for (const action of actions) {
+            await action()
+          }
+        })
+      )
+
+      // All the remaining groups provided by LDAP can be removed from XO since
+      // they don't exist in the LDAP directory any more
+      await Promise.all(
+        xoGroups
+          .filter(group => group.provider === 'ldap')
+          .map(group => this._xo.deleteGroup(group.id))
+      )
+
+      return groups
+    } finally {
+      await client.unbind()
+    }
+  }
+
+  async _synchronizeUser(ldapUsername, ldapUser) {
+    const { merge, idAttribute } = this._usersConfig
+
+    const ldapId = ldapUser[idAttribute]
+    const xoUsers = await this._xo.getAllUsers()
+    let xoUser = xoUsers.find(
+      xoUser => xoUser.authProviders?.ldap?.id === ldapId
+    )
+    const conflictingXoUser = xoUsers.find(
+      user => user.email === ldapUsername && user.id !== xoUser?.id
+    )
+
+    if (conflictingXoUser !== undefined) {
+      if (xoUser === undefined || merge) {
+        xoUser = conflictingXoUser
+      } else {
+        throw new Error('XO user with same username already exists')
+      }
+    }
+
+    if (xoUser === undefined) {
+      return this._xo.createUser({
+        name: ldapUsername,
+        authProviders: { ldap: { id: ldapId } },
+      })
+    }
+
+    await this._xo.updateUser(xoUser.id, {
+      name: ldapUsername,
```

Comment:
I'm conflicted about this.

In this case I believe it makes sense but in general, there can multiple providers attached to a user, and providers should not dictate the name of the user.

Reply:
As a work-around, we can first assert there are no other providers for this user.